### ISSUE TITLE
`quantity_input` decorator is un-typed

### ISSUE THREAD

**Issue by dstansby** [astropy/astropy#14110]
### Description
I'm trying to add typing to some code I use with `Quantity` input.ouput. Using the `u.quantity_input` decorator errors with `mypy` though, complaining that the decorator is untyped.

### Expected behavior
No complaint from `mypy`

### Actual behavior
`mypy` errors with
```python
test.py:3: error: Untyped decorator makes function "speed" untyped
Found 1 error in 1 file (checked 1 source file)
```

### Steps to Reproduce

1. Save the following code to file
2. run `mypy --strict test.py`

```python
import astropy.units as u

@u.quantity_input
def speed(distance: u.m, time: u.s) -> u.Quantity:
    return distance / time
```

### System Details

macOS-12.6.1-arm64-arm-64bit
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29) [Clang 14.0.6 ]
Numpy 1.23.4
pyerfa 2.0.0.1
astropy 5.1.1
Scipy 1.9.3
Matplotlib 3.6.2
mypy 0.982 (compiled: yes)

**Comment by eerovaher** [comment]
See https://github.com/astropy/astropy/issues/14134#issuecomment-1341247450

**Comment by mhvk** [comment]
For `quantity_input`, @nstarman had a nice PR but it became a bit messy; see https://github.com/astropy/astropy/pull/10655 and hence that was closed. So far, what has been implemented is the part that one can do `Quantity[unit]`  (https://github.com/astropy/astropy/pull/10662)

**Comment by nstarman** [comment]
Yeah. Typing ``quantity_input`` wouldn't actually be all that hard. I would welcome a PR. The underlying issues are that we are hesitant to add a lot of typing before an APE.

**Comment by Cadair** [comment]
I will just reiterate from #14134 that `quantity_inputs` existing use of function annotations as not-type-hints is perfectly legitimate and should remain supported even if it learns how to read valid type hints in addition.
